TestImageOperationsGolden: arm64, ppc64le, s390x generate slightly different PNGs due to "fused multiply and add" (FMA) instruction

On arm64, ppc64el (ppc64le) and s390x platforms, TestImageOperationsGolden reveals that Hugo (or github.com/disintegration/gift or an underlying library it uses) generates image with different file sizes, as seen on Debian auto-builders and Ubuntu snap builders, and reproducible on e.g. Debian s390x porterbox zelenka.debian.org:

* https://buildd.debian.org/status/package.php?p=hugo
* https://launchpad.net/~gohugoio/+snap/hugo-dev

For example:

* File: gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_84b0614b9f84c94c0773ef49ae868d0b.png
* want: size: 58776
* got: size: 58791

The files that differ on s390x are as follows:

* gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_300x200_fill_gaussian_smart1_2.png (this one in particular has same file size but different content from byte 16895 on)
* gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_84b0614b9f84c94c0773ef49ae868d0b.png
* gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_9c5c204a4fc82e861344066bc8d0c7db.png
* gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_a0088abf33fdbf6be1651a71e7d4dc33.png
* gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_cfc2eacca4b2748852f953954207d615.png
* gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_d1ad299f68cb4b3e1eba2ab7633e7857.png
* gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_dd36fa3cc8ae7cf4d686caf1a171284b.png
* gohugoio8_hu7f72c00afdf7634587afaa5eff2a25b2_73538_300x200_fill_gaussian_smart1_2.png
* gohugoio8_hu7f72c00afdf7634587afaa5eff2a25b2_73538_4c320010919da2d8b63ed24818b4d8e1.png
* gohugoio8_hu7f72c00afdf7634587afaa5eff2a25b2_73538_9d4c2220235b3c2d9fa6506be571560f.png
* gohugoio8_hu7f72c00afdf7634587afaa5eff2a25b2_73538_c74bb417b961e09cf1aac2130b7b9b85.png

Further investigation is needed and pending.

### What version of Hugo are you using (`hugo version`)?

0.58.x and 0.59-DEV

### Does this issue reproduce with the latest release?

Yes.

### Sample full error output

```
--- FAIL: TestImageOperationsGolden (7.48s)
    ...
    quicktest.go:288: 
        error:
          values are not deep equal
        diff (-got +want):
            (*os.fileStat)(
          - 	&{
          - 		name: "gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_84b0614b9f84c94c0773ef49ae868d0b.png",
          - 		size: 58791,
          - 		mode: 436,
          - 		modTime: time.Time{
          - 			wall: 0x2a36c6d5,
          - 			ext:  63705700464,
          - 			loc: &time.Location{
          - 				name:       "Local",
          - 				zone:       []time.zone{{name: "UTC"}},
          - 				tx:         []time.zoneTrans{{when: -9223372036854775808}},
          - 				cacheStart: -9223372036854775808,
          - 				cacheEnd:   9223372036854775807,
          - 				cacheZone:  ⟪0x400053e220⟫,
          - 			},
          - 		},
          - 		sys: syscall.Stat_t{
          - 			Dev:     0x29,
          - 			Ino:     0x23341e,
          - 			Mode:    0x81b4,
          - 			Nlink:   0x01,
          - 			Uid:     0x0b88,
          - 			Gid:     0x03f1,
          - 			Size:    58791,
          - 			Blksize: 4096,
          - 			Blocks:  120,
          - 			Atim:    syscall.Timespec{Sec: 1570103664, Nsec: 708232917},
          - 			Mtim:    syscall.Timespec{Sec: 1570103664, Nsec: 708232917},
          - 			Ctim:    syscall.Timespec{Sec: 1570103664, Nsec: 708232917},
          - 		},
          - 	},
          + 	&{
          + 		name: "gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_84b0614b9f84c94c0773ef49ae868d0b.png",
          + 		size: 58776,
          + 		mode: 420,
          + 		modTime: time.Time{
          + 			wall: 0x255280f2,
          + 			ext:  63705700209,
          + 			loc: &time.Location{
          + 				name:       "Local",
          + 				zone:       []time.zone{{name: "UTC"}},
          + 				tx:         []time.zoneTrans{{when: -9223372036854775808}},
          + 				cacheStart: -9223372036854775808,
          + 				cacheEnd:   9223372036854775807,
          + 				cacheZone:  ⟪0x400053e220⟫,
          + 			},
          + 		},
          + 		sys: syscall.Stat_t{
          + 			Dev:     0x29,
          + 			Ino:     0x22d5ce,
          + 			Mode:    0x81a4,
          + 			Nlink:   0x01,
          + 			Uid:     0x0b88,
          + 			Gid:     0x03f1,
          + 			Size:    58776,
          + 			Blksize: 4096,
          + 			Blocks:  120,
          + 			Atim:    syscall.Timespec{Sec: 1570103409, Nsec: 626163954},
          + 			Mtim:    syscall.Timespec{Sec: 1570103409, Nsec: 626163954},
          + 			Ctim:    syscall.Timespec{Sec: 1570103409, Nsec: 626163954},
          + 		},
          + 	},
            )
        got:
          &os.fileStat{
              name:    "gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_84b0614b9f84c94c0773ef49ae868d0b.png",
              size:    58791,
              mode:    0x1b4,
              modTime: time.Time{
                  wall: 0x2a36c6d5,
                  ext:  63705700464,
                  loc:  &time.Location{
                      name: "Local",
                      zone: {
                          {name:"UTC", offset:0, isDST:false},
                      },
                      tx: {
                          {when:-9223372036854775808, index:0x0, isstd:false, isutc:false},
                      },
                      cacheStart: -9223372036854775808,
                      cacheEnd:   9223372036854775807,
                      cacheZone:  &time.zone{(CYCLIC REFERENCE)},
                  },
              },
              sys: syscall.Stat_t{
                  Dev:               0x29,
                  Ino:               0x23341e,
                  Mode:              0x81b4,
                  Nlink:             0x1,
                  Uid:               0xb88,
                  Gid:               0x3f1,
                  Rdev:              0x0,
                  X__pad1:           0x0,
                  Size:              58791,
                  Blksize:           4096,
                  X__pad2:           0,
                  Blocks:            120,
                  Atim:              syscall.Timespec{Sec:1570103664, Nsec:708232917},
                  Mtim:              syscall.Timespec{Sec:1570103664, Nsec:708232917},
                  Ctim:              syscall.Timespec{Sec:1570103664, Nsec:708232917},
                  X__glibc_reserved: {0, 0},
              },
          }
        want:
          &os.fileStat{
              name:    "gohugoio24_huc57dd738f4724f4b341121e66fd85555_267952_84b0614b9f84c94c0773ef49ae868d0b.png",
              size:    58776,
              mode:    0x1a4,
              modTime: time.Time{
                  wall: 0x255280f2,
                  ext:  63705700209,
                  loc:  &time.Location{
                      name: "Local",
                      zone: {
                          {name:"UTC", offset:0, isDST:false},
                      },
                      tx: {
                          {when:-9223372036854775808, index:0x0, isstd:false, isutc:false},
                      },
                      cacheStart: -9223372036854775808,
                      cacheEnd:   9223372036854775807,
                      cacheZone:  &time.zone{(CYCLIC REFERENCE)},
                  },
              },
              sys: syscall.Stat_t{
                  Dev:               0x29,
                  Ino:               0x22d5ce,
                  Mode:              0x81a4,
                  Nlink:             0x1,
                  Uid:               0xb88,
                  Gid:               0x3f1,
                  Rdev:              0x0,
                  X__pad1:           0x0,
                  Size:              58776,
                  Blksize:           4096,
                  X__pad2:           0,
                  Blocks:            120,
                  Atim:              syscall.Timespec{Sec:1570103409, Nsec:626163954},
                  Mtim:              syscall.Timespec{Sec:1570103409, Nsec:626163954},
                  Ctim:              syscall.Timespec{Sec:1570103409, Nsec:626163954},
                  X__glibc_reserved: {0, 0},
              },
          }
        stack:
          /<<PKGBUILDDIR>>/obj-aarch64-linux-gnu/src/github.com/gohugoio/hugo/resources/image_test.go:543
            c.Assert(fi1, eq, fi2)
        
FAIL
FAIL	github.com/gohugoio/hugo/resources	48.819s
```

### Archived build logs in case they go away in the future:

* http://web.archive.org/web/20191003180051/https://buildd.debian.org/status/package.php?p=hugo
* http://web.archive.org/web/20191003182821/https://launchpad.net/~gohugoio/+snap/hugo-dev

* https://launchpadlibrarian.net/444820590/buildlog_snap_ubuntu_xenial_arm64_hugo-dev_BUILDING.txt.gz
* https://launchpadlibrarian.net/444821645/buildlog_snap_ubuntu_xenial_ppc64el_hugo-dev_BUILDING.txt.gz
* https://launchpadlibrarian.net/444820127/buildlog_snap_ubuntu_xenial_s390x_hugo-dev_BUILDING.txt.gz

--- a/resources/image_test.go
+++ b/resources/image_test.go
@@ -15,6 +15,7 @@ package resources
 
 import (
 	"fmt"
+	"image"
 	"io/ioutil"
 	"math/big"
 	"math/rand"
@@ -22,6 +23,7 @@ import (
 	"path"
 	"path/filepath"
 	"regexp"
+	"runtime"
 	"strconv"
 	"sync"
 	"testing"
@@ -478,6 +480,47 @@ func BenchmarkImageExif(b *testing.B) {
 
 }
 
+// usesFMA indicates whether "fused multiply and add" (FMA) instruction is
+// used.  The command "grep FMADD go/test/codegen/floats.go" can help keep
+// the FMA-using architecture list updated.
+var usesFMA = runtime.GOARCH == "s390x" ||
+	runtime.GOARCH == "ppc64" ||
+	runtime.GOARCH == "ppc64le" ||
+	runtime.GOARCH == "arm64"
+
+// goldenEqual compares two NRGBA images.  It is used in golden tests only.
+// A small tolerance is allowed on architectures using "fused multiply and add"
+// (FMA) instruction to accommodate for floating-point rounding differences
+// with control golden images that were generated on amd64 architecture.
+// See https://golang.org/ref/spec#Floating_point_operators
+// and https://github.com/gohugoio/hugo/issues/6387 for more information.
+//
+// Borrowed from https://github.com/disintegration/gift/blob/a999ff8d5226e5ab14b64a94fca07c4ac3f357cf/gift_test.go#L598-L625
+// Copyright (c) 2014-2019 Grigory Dryapak
+// Licensed under the MIT License.
+func goldenEqual(img1, img2 *image.NRGBA) bool {
+	maxDiff := 0
+	if usesFMA {
+		maxDiff = 1
+	}
+	if !img1.Rect.Eq(img2.Rect) {
+		return false
+	}
+	if len(img1.Pix) != len(img2.Pix) {
+		return false
+	}
+	for i := 0; i < len(img1.Pix); i++ {
+		diff := int(img1.Pix[i]) - int(img2.Pix[i])
+		if diff < 0 {
+			diff = -diff
+		}
+		if diff > maxDiff {
+			return false
+		}
+	}
+	return true
+}
+
 func TestImageOperationsGolden(t *testing.T) {
 	c := qt.New(t)
 	c.Parallel()
@@ -580,21 +623,51 @@ func TestImageOperationsGolden(t *testing.T) {
 		}
 		fi2 := dirinfos2[i]
 		c.Assert(fi1.Name(), qt.Equals, fi2.Name())
-		c.Assert(fi1, eq, fi2)
+
 		f1, err := os.Open(filepath.Join(dir1, fi1.Name()))
 		c.Assert(err, qt.IsNil)
 		f2, err := os.Open(filepath.Join(dir2, fi2.Name()))
 		c.Assert(err, qt.IsNil)
 
-		hash1, err := helpers.MD5FromReader(f1)
+		img1, _, err := image.Decode(f1)
 		c.Assert(err, qt.IsNil)
-		hash2, err := helpers.MD5FromReader(f2)
+		img2, _, err := image.Decode(f2)
 		c.Assert(err, qt.IsNil)
+
+		nrgba1 := image.NewNRGBA(img1.Bounds())
+		gift.New().Draw(nrgba1, img1)
+		nrgba2 := image.NewNRGBA(img2.Bounds())
+		gift.New().Draw(nrgba2, img2)
+
+		if !goldenEqual(nrgba1, nrgba2) {
+			switch fi1.Name() {
+			case "gohugoio8_hu7f72c00afdf7634587afaa5eff2a25b2_73538_4c320010919da2d8b63ed24818b4d8e1.png",
+				"gohugoio8_hu7f72c00afdf7634587afaa5eff2a25b2_73538_9d4c2220235b3c2d9fa6506be571560f.png",
+				"gohugoio8_hu7f72c00afdf7634587afaa5eff2a25b2_73538_c74bb417b961e09cf1aac2130b7b9b85.png":
+				c.Log("expectedly differs from golden due to dithering:", fi1.Name())
+			default:
+				t.Errorf("resulting image differs from golden: %s", fi1.Name())
+			}
+		}
+
+		if !usesFMA {
+			c.Assert(fi1, eq, fi2)
+
+			_, err = f1.Seek(0, 0)
+			c.Assert(err, qt.IsNil)
+			_, err = f2.Seek(0, 0)
+			c.Assert(err, qt.IsNil)
+
+			hash1, err := helpers.MD5FromReader(f1)
+			c.Assert(err, qt.IsNil)
+			hash2, err := helpers.MD5FromReader(f2)
+			c.Assert(err, qt.IsNil)
+
+			c.Assert(hash1, qt.Equals, hash2)
+		}
 
 		f1.Close()
 		f2.Close()
-
-		c.Assert(hash1, qt.Equals, hash2)
 	}
 
 }

